pkg/application/service/factory/service_factory.go

```diff
@@ -0,0 +1,117 @@
+package factory
+
+import (
+	"github.com/codeready-toolchain/registration-service/pkg/application/service"
+	"github.com/codeready-toolchain/registration-service/pkg/kubeclient"
+	signup_service "github.com/codeready-toolchain/registration-service/pkg/signup/service"
+	verification_service "github.com/codeready-toolchain/registration-service/pkg/verification/service"
+	"github.com/prometheus/common/log"
+
+	servicecontext "github.com/codeready-toolchain/registration-service/pkg/application/service/context"
+	"github.com/codeready-toolchain/registration-service/pkg/configuration"
+)
+
+type serviceContextImpl struct {
+	kubeClient kubeclient.CRTClient
+	services   service.Services
+}
+
+type ServiceContextOption = func(ctx *serviceContextImpl)
+
+func CRTClientOption(kubeClient kubeclient.CRTClient) ServiceContextOption {
+	return func(ctx *serviceContextImpl) {
+		ctx.kubeClient = kubeClient
+	}
+}
+
+func (s *serviceContextImpl) CRTClient() kubeclient.CRTClient {
+	return s.kubeClient
+}
+
+func (s *serviceContextImpl) Services() service.Services {
+	return s.services
+}
+
+type ServiceFactory struct {
+	contextProducer            servicecontext.ServiceContextProducer
+	serviceContextOptions      []ServiceContextOption
+	config                     configuration.Configuration
+	verificationServiceFunc    func(opts ...verification_service.VerificationServiceOption) service.VerificationService
+	verificationServiceOptions []verification_service.VerificationServiceOption
+}
+
+func (s *ServiceFactory) defaultServiceContextProducer() servicecontext.ServiceContextProducer {
+	return func() servicecontext.ServiceContext {
+		return &serviceContextImpl{
+			services: s,
+		}
+	}
+}
+
+func (s *ServiceFactory) SignupService() service.SignupService {
+	return signup_service.NewSignupService(s.getContext(), s.config)
+}
+
+func (s *ServiceFactory) VerificationService() service.VerificationService {
+	return s.verificationServiceFunc(s.verificationServiceOptions...)
+}
+
+func (s *ServiceFactory) WithVerificationServiceOption(opt verification_service.VerificationServiceOption) {
+	s.verificationServiceOptions = append(s.verificationServiceOptions, opt)
+}
+
+// Option an option to configure the Service Factory
+type Option func(f *ServiceFactory)
+
+func WithServiceContextOptions(opts ...ServiceContextOption) func(f *ServiceFactory) {
+	return func(f *ServiceFactory) {
+		for _, opt := range opts {
+			f.serviceContextOptions = append(f.serviceContextOptions, opt)
+		}
```

Comment:
I guess that you can replace it with one simple append command:
```suggestion
		f.serviceContextOptions = append(f.serviceContextOptions, opts...)
```

Reply:
Nice suggestion, I didn't realise the append function was a variadic.